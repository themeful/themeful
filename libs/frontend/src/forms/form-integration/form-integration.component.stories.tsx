/* eslint-disable react/jsx-no-bind */
import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import { Subject } from 'rxjs'

const triggerOverlay = new Subject()
const openOverlay = (form) => {
  triggerOverlay.next(formDatas[form])
}
const groups = ['First Group', 'Second Group', 'Third Group']
const propertyTypes = [
  { key: 'color', value: 'Color' },
  { key: 'font-size', value: 'Font Size' },
  { key: 'mediaquery', value: 'MediaQuery' },
]

const formDatas = {
  styleguide: {
    form: 'styleguide',
    identifier: 'styleGuide1',
    fields: {
      name: 'StyleGuide Name',
      baseFontSize: 16,
    },
  },
  style: {
    form: 'style',
    identifier: { styleguide: 'styleGuide1', style: 'baseGroup_red' },
    groups: groups,
    propertyTypes: propertyTypes,
    fields: {
      type: 'color',
      group: 'Base Group',
      name: 'Red',
      value: '#ff0000',
    },
  },
  empty: {
    form: 'empty',
    identifier: 'styleGuide1',
    fields: {
      name: 'StyleGuide Name',
      baseFontSize: 16,
    },
  },
}

export default {
  title: 'Forms/Form Integration',
}

export const formIntegration = (args) => {
  return (
    <div>
      <div class="header">
        <h1>Form Integration</h1>
      </div>
      <div class="content">
        <div style={{ width: '80%', display: 'flex', gap: '16px' }}>
          <tf-button onClick={() => openOverlay('styleguide')}>Open Style Guide Form</tf-button>
          <tf-button onClick={() => openOverlay('style')}>Open Style Form</tf-button>
          <tf-button onClick={() => openOverlay('empty')}>Open Empty</tf-button>
        </div>
        <tf-form-integration {...args} />
      </div>
    </div>
  )
}

formIntegration.args = {
  formData$: triggerOverlay,
  onAction: action('triggered action'),
}
