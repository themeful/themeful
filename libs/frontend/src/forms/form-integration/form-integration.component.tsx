import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core'
import { FormIntegrationActions, FormIntegrations } from '@typings'
import { Subject, Subscription } from 'rxjs'

@Component({
  tag: 'tf-form-integration',
  styleUrl: 'form-integration.component.scss',
  shadow: true,
})
export class FormIntegrationComponent {
  /** FormData */
  @Prop() formData$ = new Subject<FormIntegrations>()
  @State() show = false
  @State() formData: FormIntegrations

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action: EventEmitter<FormIntegrationActions>

  private sub = new Subscription()
  private args = {}

  private close = (): void => {
    this.show = false
  }

  private onAction = ({ detail }): void => {
    console.log('action', detail)
    if (detail.action !== 'close') {
      this.action.emit({ ...detail, controller: this.formData.form })
    }

    if (detail.action !== 'open') {
      this.close()
    }
  }

  public componentWillLoad(): void {
    this.sub.add(
      this.formData$.subscribe((formData) => {
        this.formData = formData
        this.args = { formData, onAction: this.onAction }
        this.show = true
      })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  private renderForm(form: string): HTMLElement {
    switch (form) {
      case 'styleguide':
        return <tf-style-guide-form {...this.args} />
      case 'style':
        return <tf-style-form {...this.args} />
      case 'theme':
        return <tf-theme-form {...this.args} />
      case 'themeValue':
        return <tf-theme-value-form {...this.args} />
      case 'designToken':
        return <tf-design-token-form {...this.args} />
      case 'aliasTokenSelect':
        return <tf-alias-token-form {...this.args} />
      default:
        return <div>Form is not defined</div>
    }
  }

  public render(): HTMLTfFormIntegrationElement {
    return (
      <tf-overlay {...{ show: this.show, onClose: this.close }}>
        {this.show && this.renderForm(this.formData.form)}
      </tf-overlay>
    )
  }
}
