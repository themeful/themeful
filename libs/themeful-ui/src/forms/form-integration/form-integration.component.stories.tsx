/* eslint-disable react/jsx-no-bind */
import { propertySelect } from '@properties'
import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import { Subject } from 'rxjs'

const triggerOverlay = new Subject()
const openOverlay = (form) => {
  triggerOverlay.next(formDatas[form])
}
const groups = ['First Group', 'Second Group', 'Third Group']

const styles = {
  global_base_black: {
    name: 'Black',
    type: 'color',
    group: 'base',
    value: '#333333',
    global: true,
  },
  global_base_light: {
    name: 'Light',
    type: 'color',
    group: 'base',
    value: '#ffffff',
    global: true,
  },
  global_gray_100: {
    name: '100',
    type: 'color',
    group: 'gray',
    value: 'rgba(0, 0, 0, 0.75)',
    global: true,
  },
  global_gray_200: {
    name: '200',
    type: 'color',
    group: 'gray',
    value: 'rgba(0, 0, 0, 0.60)',
    global: true,
  },
  global_gray_300: {
    name: '300',
    type: 'color',
    group: 'gray',
    value: 'rgba(0, 0, 0, 0.45)',
    global: true,
  },
  global_gray_400: {
    name: '400',
    type: 'color',
    group: 'gray',
    value: 'rgba(0, 0, 0, 0.30)',
    global: true,
  },
  global_gray_500: {
    name: '500',
    type: 'color',
    group: 'gray',
    value: 'rgba(0, 0, 0, 0.15)',
    global: true,
  },
  styleGuide1_action_primary: {
    name: 'Primary',
    type: 'color',
    group: 'action',
    value: '#31ed31',
    global: false,
  },
  styleGuide1_action_secondary: {
    name: 'Secondary',
    type: 'color',
    group: 'action',
    value: '#2ec22e',
    global: false,
  },
  styleGuide1_action_tertiary: {
    name: 'Tertiary',
    type: 'color',
    group: 'action',
    value: '#1e961e',
    global: false,
  },
  styleGuide1_brand_primary: {
    type: 'color',
    group: 'brand',
    name: 'Primary',
    value: '#ff0000',
    global: false,
  },
  styleGuide1_brand_secondary: {
    name: 'Secondary',
    type: 'color',
    group: 'brand',
    value: '#d42828',
    global: false,
  },
  styleGuide1_brand_tertiary: {
    name: 'Tertiary',
    type: 'color',
    group: 'brand',
    value: '#af2323',
    global: false,
  },
}

const formDatas = {
  styleguideDuplicate: {
    form: 'styleguideDuplicate',
    identifier: 'styleGuide1',
    fields: {
      name: 'Style Guide 1',
    },
  },
  styleguide: {
    form: 'styleguide',
    identifier: 'styleGuide1',
    fields: {
      name: 'StyleGuide Name',
      baseFontSize: 16,
    },
  },
  style: {
    form: 'style',
    identifier: { styleguide: 'styleGuide1', style: 'baseGroup_red' },
    groups,
    propertyTypes: propertySelect,
    fields: {
      type: 'color',
      group: 'Base Group',
      name: 'Red',
      value: '#ff0000',
    },
  },
  designToken: {
    form: 'designToken',
    identifier: 'dtSomeTokenName',
    groups,
    propertyTypes: propertySelect,
    fields: {
      name: 'Some Token Name',
      group: 'Base',
      type: 'color',
      description: 'Some good description',
    },
  },
  themeValue: {
    form: 'themeValue',
    identifier: {
      designToken: 'dtActionBg',
      theme: 'styleGuide1_light',
      media: 'default',
    },
    styles,
    type: 'color',
    medias: [
      {
        key: 'default',
        value: 'Default',
      },
      {
        key: 'global_mediaQuery_aboveSmallMobile',
        value: 'Above Small Mobile',
      },
      {
        key: 'global_mediaQuery_aboveMobile',
        value: 'Above Mobile',
      },
      {
        key: 'global_mediaQuery_aboveTablet',
        value: 'Above Tablet',
      },
      {
        key: 'global_mediaQuery_aboveDesktop',
        value: 'Above Desktop',
      },
    ],
    fields: {
      media: 'default',
      style: 'styleGuide1_action_primary',
    },
  },
  theme: {
    form: 'theme',
    identifier: 'styleGuide1_dark',
    styleGuides: [
      { key: 'styleGuide1', value: 'Style Guide 1' },
      { key: 'styleGuide2', value: 'Style Guide 2' },
    ],
    fields: {
      name: 'Dark',
      styleGuide: 'styleGuide1',
    },
  },
  aliasTokenSelect: {
    form: 'aliasTokenSelect',
    identifier: 'dtActionBg',
    aliasTokens: ['atCardBackground'],
    fields: {
      selected: ['atButtonBackground'],
    },
  },
  empty: {
    form: 'empty',
  },
}

export default {
  title: 'Form Integration/All Forms',
}

export const allForms = (args) => {
  return (
    <div>
      <div class="header">
        <h1>Form Integration</h1>
      </div>
      <div class="content">
        <div
          style={{
            width: '100%',
            display: 'grid',
            gap: '16px',
            justifyContent: 'center',
            justifyItems: 'center',
            gridTemplateColumns: '1fr 1fr 1fr',
          }}
        >
          <tf-button onClick={() => openOverlay('styleguide')}>Open Style Guide Form</tf-button>
          <tf-button onClick={() => openOverlay('style')}>Open Style Form</tf-button>
          <tf-button onClick={() => openOverlay('theme')}>Open Theme Form</tf-button>
          <tf-button onClick={() => openOverlay('themeValue')}>Open Theme Value Form</tf-button>
          <tf-button onClick={() => openOverlay('designToken')}>Open Design Token Form</tf-button>
          <tf-button onClick={() => openOverlay('aliasTokenSelect')}>
            Open Alias Token Form
          </tf-button>
          <tf-button onClick={() => openOverlay('styleguideDuplicate')}>
            Open Style Guide Duplicate Form
          </tf-button>
          <tf-button onClick={() => openOverlay('empty')}>Open Empty</tf-button>
        </div>
        <tf-form-integration {...args} />
      </div>
    </div>
  )
}

allForms.args = {
  formData$: triggerOverlay,
  onAction: action('triggered action'),
}
