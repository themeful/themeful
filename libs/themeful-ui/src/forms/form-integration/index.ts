export * from './form-integration.component'
