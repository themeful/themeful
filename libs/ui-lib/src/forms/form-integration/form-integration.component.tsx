import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core'
import { FormIntegrationActions, FormIntegrations } from '@typings'
import { Subject, Subscription } from 'rxjs'
import '../../components/overlay'
import '../../forms/alias-token-form'
import '../../forms/design-token-form'
import '../../forms/design-token-split-form'
import '../../forms/style-form'
import '../../forms/style-guide-duplicate-form'
import '../../forms/style-guide-form'
import '../../forms/theme-duplicate-form'
import '../../forms/theme-form'
import '../../forms/theme-value-form'

@Component({
  tag: 'tf-form-integration',
  styleUrl: 'form-integration.component.scss',
  shadow: true,
})
export class FormIntegrationComponent {
  /** FormData */
  @Prop() formData$ = new Subject<FormIntegrations>()
  @State() show = false
  @State() formData: FormIntegrations

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action: EventEmitter<FormIntegrationActions>

  private sub = new Subscription()
  private args = {}

  private close = (): void => {
    this.show = false
  }

  private onAction = ({ detail }): void => {
    if (detail.action !== 'close') {
      this.action.emit({ ...detail, controller: detail.controller || this.formData.form })
    }

    if (detail.action !== 'open') {
      this.close()
    }
  }

  public componentWillLoad(): void {
    this.sub.add(
      this.formData$.subscribe((formData) => {
        this.formData = formData
        this.args = { formData, onAction: this.onAction }
        this.show = true
      })
    )
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  private renderForm(form: string): HTMLElement {
    switch (form) {
      case 'styleguide':
        return <tf-style-guide-form {...this.args} />
      case 'styleguideDuplicate':
        return <tf-style-guide-duplicate-form {...this.args} />
      case 'style':
        return <tf-style-form {...this.args} />
      case 'theme':
        return <tf-theme-form {...this.args} />
      case 'themeDuplicate':
        return <tf-theme-duplicate-form {...this.args} />
      case 'themeValue':
        return <tf-theme-value-form {...this.args} />
      case 'designToken':
        return <tf-design-token-form {...this.args} />
      case 'designTokenSplit':
        return <tf-design-token-split-form {...this.args} />
      case 'aliasTokenSelect':
        return <tf-alias-token-form {...this.args} />
      default:
        return <div>Form is not defined</div>
    }
  }

  public render(): HTMLTfFormIntegrationElement {
    return (
      <tf-overlay {...{ show: this.show, onClose: this.close }}>
        {this.show && this.renderForm(this.formData.form)}
      </tf-overlay>
    )
  }
}
