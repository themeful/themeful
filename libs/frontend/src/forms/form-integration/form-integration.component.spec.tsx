import { newSpecPage } from '@stencil/core/testing'
import { FormIntegrationComponent } from './form-integration.component'

xdescribe('FormIntegrationComponent', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [FormIntegrationComponent],
      html: '<tf-form-integration></tf-form-integration>',
    })
    expect(root).toEqualHtml(`
      <tf-form-integration>
        <mock:shadow-root>
          <div>
            Hello, World! I'm
          </div>
        </mock:shadow-root>
      </tf-form-integration>
    `)
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [FormIntegrationComponent],
      html: `<tf-form-integration first="Stencil" last="'Don't call me a framework' JS"></tf-form-integration>`,
    })
    expect(root).toEqualHtml(`
      <tf-form-integration first="Stencil" last="'Don't call me a framework' JS">
        <mock:shadow-root>
          <div>
            Hello, World! I'm Stencil 'Don't call me a framework' JS
          </div>
        </mock:shadow-root>
      </tf-form-integration>
    `)
  })
})
