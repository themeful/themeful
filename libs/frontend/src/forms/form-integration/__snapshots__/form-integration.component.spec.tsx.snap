// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormIntegrationComponent renders 1`] = `
<tf-form-integration>
  <mock:shadow-root>
    <tf-overlay>
      <mock:shadow-root></mock:shadow-root>
    </tf-overlay>
  </mock:shadow-root>
</tf-form-integration>
`;
